userservice/src/main/java/com/infodation/userservice/services/UserServiceImpl.java

```diff
@@ -7,30 +7,44 @@
 import com.infodation.userservice.services.iservice.IUserService;
 import org.springframework.data.domain.Page;
 import org.springframework.data.domain.Pageable;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.scheduling.annotation.EnableAsync;
 import org.springframework.stereotype.Service;
 
 import java.time.LocalDateTime;
+import java.util.List;
 import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
 
 @Service
 public class UserServiceImpl implements IUserService {
-
     private final UserRepository userRepository;
+    public UserServiceImpl(UserRepository userRepository) {
+        this.userRepository = userRepository;
+    }
+
+    @Async
+    public CompletableFuture<Void> bulkEditUsers(List<UpdateUserDTO> usersDTO) {
+        for (UpdateUserDTO userDTO : usersDTO) {
+            User userToUpdate = userRepository.findByUserId(userDTO.getUserId()).orElse(null);
+            if (userToUpdate != null) {
+                userToUpdate.setFirstName(userDTO.getFirstName());
+                userToUpdate.setLastName(userDTO.getLastName());
+                userToUpdate.setEmail(userDTO.getEmail());
+                userToUpdate.setSex(userDTO.getSex());
+                userToUpdate.setUpdatedAt(LocalDateTime.now());
+                userRepository.save(userToUpdate);
+            }
+        }
```

Comment:
These lines of code will make many connections into database due ot `userRepository.save(userToUpdate);` for each entities you find.
I suggest using `saveAll ` method from JPA.

This means you need to save the list of to be updated users and do call `userRepository.saveAll(usersToUpdate);` at once